### PR TITLE
Implement Base.setindex with CartesianIndex

### PR DESCRIPTION
**Pull request by eschnett**
Closes https://github.com/JuliaArrays/StaticArrays.jl/issues/1033.

### CODE REVIEW

File: src/deque.jl [hunk 1]
```diff
@@ -196,3 +196,4 @@ end
 
 # TODO proper multidimension boundscheck
 @propagate_inbounds setindex(a::StaticArray, x, inds::Int...) = setindex(a, x, LinearIndices(a)[inds...])
+@propagate_inbounds setindex(a::StaticArray, x, I::CartesianIndex) = setindex(a, x, I.I...)
```
Comment (jw3126) [inline src/deque.jl:199]: Not a big deal, but I guess `I.I` is an implementation detail?
```suggestion
@propagate_inbounds setindex(a::StaticArray, x, I::CartesianIndex) = setindex(a, x, Tuple(I)...)
```

Comment (jw3126) [inline src/deque.jl:199]: I just looked at  `?CartesianIndex` and it says:
> One can freely mix integer and CartesianIndex indices;
  for example, A[Ipre, i, Ipost] (where Ipre and Ipost are CartesianIndex indices and i is an Int) can be a useful expression when writing algorithms that work along a single dimension of an array of
  arbitrary dimensionality.


```suggestion
@propagate_inbounds setindex(a::StaticArray, x, inds::Union{Integer, CartesianIndex}...) = setindex(a, x, LinearIndices(a)[inds...])
```

Comment (eschnett) [reply src/deque.jl]: This is now irrelevant with the change below.

File: src/deque.jl [hunk 2]
```diff
@@ -195,4 +195,4 @@ julia> setindex(@SMatrix[2 4; 6 8], 1, 2)
 end
 
 # TODO proper multidimension boundscheck
-@propagate_inbounds setindex(a::StaticArray, x, inds::Int...) = setindex(a, x, LinearIndices(a)[inds...])
+@propagate_inbounds setindex(a::StaticArray, x, inds::Union{Integer, CartesianIndex}...) = setindex(a, x, LinearIndices(a)[inds...])
```
Comment (johnnychen94) [inline src/deque.jl:198]: I'm new to StaticArrays internal so I might be wrong.
Is there any reason why we don't use `to_indices` here?

```julia
@propagate_inbounds setindex(a::StaticArray, x, inds...) = setindex(a, x, LinearIndices(a)[to_indices(a, inds)...])
```

This supports the hybrid of many cases, e.g., `setindex(a, x, 1, CartesianIndex(2,))`.

Comment (eschnett) [reply src/deque.jl]: I'm trying this.

Comment (jw3126) [reply src/deque.jl]: What is the advantage of inserting `to_indices`? Won't
`LinearIndices(a)[inds...]`
call `to_indices(LinearIndices(a), inds)` anyway?

Comment (johnnychen94) [reply src/deque.jl]: Ooh you're right -- my first draft version doesn't use `LinearIndices` and it triggers recursive self-calling and OverflowError so I blindly added back the `LinearIndices` 🙈

Comment (jw3126) [reply src/deque.jl]: Yes, we should remove the `::Union{Integer, CartesianIndex}` annotation.

Comment (eschnett) [reply src/deque.jl]: I updated the code. Please have a look at the current version
```Julia
@propagate_inbounds setindex(a::StaticArray, x, inds...) = setindex(a, x, LinearIndices(a)[inds...])
```
and tell me if that should be improved.

**Comment by eschnett** [comment]
I think this is ready to be applied.

File: src/deque.jl (hunk 2 above)
Comment (jw3126) [reply src/deque.jl]: I think that looks good.

**Review by hyrodium — APPROVED** [review]
LGTM!

**Comment by eschnett** [comment]
If we make this a minor release then I can use this feature right away...

**Pull request merged**